src/Drips.sol

```diff
@@ -545,33 +544,35 @@ abstract contract Drips {
         uint256 assetId,
         DripsReceiver[] memory currReceivers,
         int128 balanceDelta,
-        DripsReceiver[] memory newReceivers
+        DripsReceiver[] memory newReceivers,
+        uint32 maxEndTip1,
+        uint32 maxEndTip2
     ) internal returns (uint128 newBalance, int128 realBalanceDelta) {
         DripsState storage state = _dripsStorage().states[assetId][userId];
-        bytes32 currDripsHash = _hashDrips(currReceivers);
-        require(currDripsHash == state.dripsHash, "Invalid current drips list");
+        require(_hashDrips(currReceivers) == state.dripsHash, "Invalid current drips list");
         uint32 lastUpdate = state.updateTime;
-        uint32 currMaxEnd = state.maxEnd;
+        uint32 newMaxEnd;
         {
-            uint128 lastBalance = state.balance;
-            uint128 currBalance =
-                _balanceAt(lastBalance, lastUpdate, currMaxEnd, currReceivers, _currTimestamp());
-            int136 balance = int128(currBalance) + int136(balanceDelta);
-            if (balance < 0) {
-                balance = 0;
+            uint32 currMaxEnd = state.maxEnd;
+            int128 currBalance = int128(
+                _balanceAt(state.balance, lastUpdate, currMaxEnd, currReceivers, _currTimestamp())
+            );
+            realBalanceDelta = balanceDelta;
+            // Cap `realBalanceDelta` at withdrawal of the entire `currBalance`
+            if (realBalanceDelta < -currBalance) {
+                realBalanceDelta = -currBalance;
             }
-            newBalance = uint128(uint136(balance));
-            realBalanceDelta = int128(balance - int128(currBalance));
+            newBalance = uint128(currBalance + realBalanceDelta);
```

Comment:
I think the flow would be easier to understand if it would be:

```solidity
 // Cap `realBalanceDelta` at withdrawal of the entire `currBalance`
 if ((currBalance + balanceDelta) < 0 ) {
     realBalanceDelta = -currBalance;
     newBalance = 0;
  } else {
     realBalanceDelta = balanceDelta;
     newBalance = uint128(currBalance + realBalanceDelta);
 }
```
However, I see how the current implementation is a bit more optimized in terms of gas.  :+1: